higgstautau/lephad/filters.py

```diff
@@ -0,0 +1,403 @@
+from rootpy.tree.filtering import EventFilter
+
+from math import *
+import math
+
+from .. import utils
+from ..units import GeV
+from .. import datasets
+from . import track_counting
+from .. import tauid
+from ..tauid import IDLOOSE, IDMEDIUM, IDTIGHT
+from . import log; log = log[__name__]
+
+class LeptonSelection(EventFilter):
+    def __init__(self,min_leps,**kwargs):
+        super(LeptonSelection,self).__init__(**kwargs)
+        self.min_leps = min_leps
+        #require lepton candidates
+    def passes(self, event):
+        return len(event.electrons) + len(event.muons) >=self.min_leps
+
+class LepPT(EventFilter):
+    def __init__(self, min_leps, el_thresh=15*GeV, mu_thresh=10 * GeV, **kwargs):
+        self.min_leps = min_leps
+        self.el_thresh = el_thresh
+        self.mu_thresh = mu_thresh
+        super(LepPT, self).__init__(**kwargs)
+    def passes(self, event):
+        event.electrons.select(lambda electron: electron.cl_E/cosh(electron.tracketa) > self.el_thresh)
+        event.muons.select(lambda muon: muon.pt > self.mu_thresh)
+        return len(event.electrons) + len(event.muons) >= self.min_leps
+
+class LepEta(EventFilter):
+    def __init__(self, min_leps, **kwargs):
+        self.min_leps = min_leps
+        super(LepEta, self).__init__(**kwargs)
+    def passes(self, event):                                                                                         
+        # both calo and leading track eta within 2.47                                                                
+        event.electrons.select(lambda electron: 
+            abs(electron.cl_eta) < 1.37 or
+            1.52 < abs(electron.cl_eta) < 2.47)
+        event.muons.select(lambda muon: abs(muon.eta) < 2.5)
+        return len(event.electrons) + len(event.muons) >= self.min_leps
+
+class LepHad(EventFilter):
+    def __init__(self, **kwargs):
+        super(LepHad, self).__init__(**kwargs)
+    def passes(self, event):
+        #make sure it's only one had and one lep
+        #sort in decending order by pT
+        event.electrons.sort(key=lambda electron: electron.cl_E/cosh(electron.tracketa), reverse=True)
+        event.muons.sort(key=lambda muon: muon.pt, reverse=True)
+        event.taus.sort(key=lambda tau: tau.pt, reverse=True)
+        if len(event.taus) != 1: 
+            return False
+        if len(event.electrons) == 1:
```

Comment:
With this selection you will keep events with one hadronic tau and *at least* a muon or an electron. Is it intended ? It seems to be in contradiction with your comment on line 49.
Also you don't need to sort the containers do you ?